File: HSE/ADS_contests/Graphs_2/Bellman_Ford2.py
def solve(n):
    INF = float('inf')
    dist = [INF] * (n + 1)
    dist[1] = 0  
    
    def weight(i, j):
        return (179 * i + 719 * j) % 1000 - 500
    
    for i in range(1, n):
        for j in range(i + 1, n + 1):
            if dist[i] != INF and dist[j] > dist[i] + weight(i, j):
                dist[j] = dist[i] + weight(i, j)
    
    return dist[n]

if __name__ == "__main__":
    n = int(input())
    print(solve(n))

File: HSE/ADS_contests/Graphs_2/Bellman_Ford2.cpp
#include <iostream>
#include <vector>
#include <limits>

int solve(int n) {
    const long long INF = std::numeric_limits<long long>::max();
    std::vector<long long> dist(n + 1, INF);
    dist[1] = 0;
    
    auto weight = [](int i, int j) -> int {
        return (179 * i + 719 * j) % 1000 - 500;
    };
    
    for (int i = 1; i < n; ++i) {
        for (int j = i + 1; j <= n; ++j) {
            if (dist[i] != INF && dist[i] + weight(i, j) < dist[j]) {
                dist[j] = dist[i] + weight(i, j);
            }
        }
    }
    
    return dist[n];
}

int main() {
    int n;
    std::cin >> n;
    std::cout << solve(n) << std::endl;
    return 0;
}